### DISCUSSION TITLE
Pre-rendered PCM data does not work as expected

### DISCUSSION THREAD

**Discussion by robots4life** [katspaugh/wavesurfer.js#2752] [Q&A]
I have created the PCM data correctly per https://wavesurfer-js.org/api/class/src/wavesurfer.js~WaveSurfer.html#instance-method-exportPCM and also checked this issue `https://github.com/wavesurfer-js/wavesurfer.js/issues/2439` as well as this comment `https://github.com/wavesurfer-js/wavesurfer.js/issues/2557#issuecomment-1199946658`.

This PCM data is then stored in an array. Here is a REPL of how I create the PCM data.
https://svelte.dev/repl/f4be2c6a630a4b66833c8fe8b06ab9e5?version=3.57.0

### wavesurfer.js version(s):
"wavesurfer.js": "^6.6.2"

### Browser and operating system version(s):
Firefox 111.0.1 (64-bit), Linux

### Code needed to reproduce the issue:
**https://svelte.dev/repl/7cee1ab9dce6422eb7b3059433948745?version=3.57.0**

### (Optional) Steps in the UI to reproduce the issue:
Once the REPL loads there are 2 ways to see that the pre-rendered PCM data does not seem to work with wavesurfer.
1. After the REPL loads click "+ 5 Seconds" and see the cursor jump to the end of the track. This means that the PCM data for the track was somehow not loaded correctly.

2. Reload the REPL, this time click "PLAY", you will notice a considerable delay before the track is played. I assume this delay comes from re-loading the PCM data. When I load longer tracks the delay before playing the track becomes 4+ seconds or more.

I would love to be able to use the generated PCM data and load it into wavesurfer correctly. Unfortunately there is no error message that comes up, `wavesurfer.load(audio, peaks);` seems to work. Kindly please do let me know if I have missed something perhaps, thank you.

**Comment by prakharpbuf** [comment]
@katspaugh I have been looking into this and it seems like just calling `load()` instead of` this.tmpEvents.push(this.once('interaction', load));` at wavesurfer.js:1503 fixes this because `this.buffer` is empty when we provide the peaks array with WebAudio backend.

This thing works fine with MediaElement.

**Reply by prakharpbuf** [reply]
it fills `this.buffer` when we first call `play()` `seekTo()` or something similar. I guess that filling of `this.buffer` when we have an interaction with the audio is the source of this problem.

**Reply by katspaugh** [reply]
Interesting. Looks like this was added by me in #734, and the reasoning was to delay the loading of the audio until necessary.
As the issue linked there says: "The audio is only downloaded in the event of a seekTo or play action."

In any case, I think it makes the most sense to use the MediaElement with pre-decoded peaks. Otherwise, with Web Audio, it will still have to download and decode the audio in order to play it as a BufferSourceNode.

So maybe @robots4life could just enable `backend: "MediaElement"`.

**Reply by robots4life** [reply]
Thank you. Given this, I will **have** to enable `backend: "MediaElement"` so that I can use it with previously created PCM data, in order to avoid decoding the audio on every load.

1.
https://wavesurfer-js.org/docs/methods.html
https://wavesurfer-js.org/api/class/src/mediaelement.js~MediaElement.html#instance-method-load
Now that I have to use `MediaElement` could you kindly tell me a bit more about `Optional preload parameter [none|metadata|auto], passed to the [Audio element](https://developer.mozilla.org/en/docs/Web/HTML/Element/audio) if using backend MediaElement.` Can I preload the PCM data somehow and hence increase preformance of the loading process, is that was this parameter is about ? 

2.
While I was integrating this in a current project I also noticed that the waveform from the PCM data differs quite a bit from the waveform that is created on load. See this REPL to see the difference in the created waveforms from  the identical track.
https://svelte.dev/repl/17ad33e46be343c9964f68e52a67ddd2?version=3.57.0
Do you know why this is the case ?

This was also a bit confusing when storing the PCM data on the server. I thought the data format was off or something else not right since the waveform I could see on load would not be the one that shows when using the PCM data.

**Reply by robots4life** [reply]
![Screenshot_20230330_173658](https://user-images.githubusercontent.com/34193243/228889578-8a3c2ea9-e817-4291-9f33-76b5d3a82ec4.png)

**Reply by robots4life** [reply]
Disabling `barWidth` and resizing the REPL inner viewport will also create a pretty random waveform on `load` while the waveform from the PCM data stays the same. This could possibly be related to https://github.com/wavesurfer-js/wavesurfer.js/issues/2439.

**Reply by prakharpbuf** [reply]
The preload parameter is something that wavesurfer passes to the HTMLMediaElement. It's not really a wavesurfer parameter. Read more about the preload option [here](https://clubmate.fi/what-does-the-preload-attribute-in-the-video-element-do). This link explains the preload option for HTML video element, but similar concepts apply to the audio element too.

Rendering the waveform with PCM data is already optimized. Notice how when you refresh the [REPL](https://svelte.dev/repl/17ad33e46be343c9964f68e52a67ddd2?version=3.57.0) the waveform with PCM data loads much faster than the one without it. This is also partly the reason why the waveforms for PCM data and load somewhat differ.

**Reply by robots4life** [reply]
@prakharpbuf thank you very much.

**Reply by robots4life** [reply]
@prakharpbuf @katspaugh Unfortunately, when using `backend: "MediaElement"` it turns out that the created PCM data is an empty array. :eyes:

**[https://svelte.dev/repl/57ee0d64d64849249de11bf114354821?version=3.58.0](https://svelte.dev/repl/57ee0d64d64849249de11bf114354821?version=3.58.0)**

This means I would have to check if a track has PCM data in the database and conditionally either load `wavesurfer` per default backend, that is WebAudio, if no PCM data is present, in order to be able to use `exportPCM` correctly to create PCM data.

But on the other hand if PCM data is present in the database instantiate `wavesurfer` with `MediaElement` so that the created PCM data can be read and loaded by `wavesurfer`.

**Reply by katspaugh** [reply]
exportPCM should work regardless of the backend. In fact, in your repl I see it logged in the console.
Edit: reloaded, and it's an empty array now. Looks like a timing bug with the ready event. If you add a small timeout before exporting, it works. Using `waveform-ready` instead of `ready` should probably also help.

From the docs:
<img width="876" alt="Screenshot 2023-03-31 at 11 48 36" src="https://user-images.githubusercontent.com/381895/229087228-b8d3795f-f9f0-48c6-bd0c-69b95bb2bdbf.png">

I the new wavesurfer.js version, there'll be a single `decode` event so this should be less confusing.

**Reply by robots4life** [reply]
@prakharpbuf @katspaugh Would you know how I can get rid of these warnings ?

1.
```js
webaudio.js:76 The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page. https://goo.gl/7K7WLu
```
![Screenshot_20230403_161426](https://user-images.githubusercontent.com/34193243/229536179-539393e4-ec57-4df8-86b4-76cda9487606.png)

2.
```js
webaudio.js:238 [Deprecation] The ScriptProcessorNode is deprecated. Use AudioWorkletNode instead. (https://bit.ly/audio-worklet)
```

![Screenshot_20230403_161626](https://user-images.githubusercontent.com/34193243/229536731-c589ab11-21ee-4ed0-aecb-9efc6b57fdee.png)

**Reply by prakharpbuf** [reply]
1. #1531 
2. #1281

**Reply by katspaugh** [reply]
Thanks @robots4life and @prakharpbuf!
I've reopened the issue about the audio context, and we need to make a new release with the script node fix.

**Reply by robots4life** [reply]
> The preload parameter is something that wavesurfer passes to the HTMLMediaElement. It's not really a wavesurfer parameter. Read more about the preload option [here](https://clubmate.fi/what-does-the-preload-attribute-in-the-video-element-do). This link explains the preload option for HTML video element, but similar concepts apply to the audio element too.
> 
> Rendering the waveform with PCM data is already optimized. Notice how when you refresh the [REPL](https://svelte.dev/repl/17ad33e46be343c9964f68e52a67ddd2?version=3.57.0) the waveform with PCM data loads much faster than the one without it. This is also partly the reason why the waveforms for PCM data and load somewhat differ.

https://github.com/wavesurfer-js/wavesurfer.js/discussions/2752#discussioncomment-5479593

@prakharpbuf  Passing the preload attribute "none" in my case increased web performance incredibly to pass PageSpeed 100/100 mobile & desktop for a page with 9 huge audio files on them, highly invaluable article and info you linked to, thanks so much.

**Reply by danielmat3** [reply]
@robots4life Hi Robert, how can I contact you? I need your help

**Comment by robots4life** [comment]
> @robots4life Hi Robert, how can I contact you? I need your help

Sure, how could I help you ?

**Reply by danielmat3** [reply]
Hey Robert!

I'm searching for a developer who can help me with WaveSurfer integration on my product pages for MP3 audio files. I want to optimize performance by setting the preload attribute to "none" for better PageSpeed scores and smoother customer experience.

I'm currently using the Wavesurfer WP plugin, but it doesn't allow me to set up preload options along with peaks.
https://wordpress.com/plugins/wavesurfer-wp

If you can assist with this, drop your email below so we can chat!

Thanks a bunch!

**Comment by robots4life** [comment]
> Hey Robert!
> 
> I'm searching for a developer who can help me with WaveSurfer integration on my product pages for MP3 audio files. I want to optimize performance by setting the preload attribute to "none" for better PageSpeed scores and smoother customer experience.
> 
> Check out my website: [unwav.com/product/groove-premium-tech-house-sample-pack/](https://unwav.com/product/groove-premium-tech-house-sample-pack/)
> 
> I'm currently using the Wavesurfer WP plugin, but it doesn't allow me to set up preload options along with peaks. https://wordpress.com/plugins/wavesurfer-wp
> 
> If you can assist with this, drop your email below so we can chat!
> 
> Thanks a bunch!

How can I contact you, let me know a suited email for your business, I could then be in touch, thank you.

**Reply by danielmat3** [reply]
You can contact me here:
business@unwav.com

**Reply by danielmat3** [reply]
Hi Robert,
Could you please check if you received my emails? I haven't heard back from you since 4 days ago.

**Comment by robots4life** [comment]
> Hi Robert, Could you please check if you received my emails? I haven't heard back from you since 4 days ago.

I sent you a message as a reply to your message from Fr 8th of March 24, since then I have not received any of your messages. I have just emailed you again, should work.

Regards
Robert

**Reply by danielmat3** [reply]
That's weird, Robert. I believe you missed my previous five emails that I've sent you. I noticed that something was off.

Is there any other email address, like Google Mail, that I can use to contact you?

I'm not sure why you didn't receive my emails. I checked my emailing system, and everything seems fine on my end.
